File: services/gateway/src/proxy.js
'use strict';

const { CONFIG } = require('@ai-fleet/shared/config');
const log = require('@ai-fleet/shared/logger');
const { idTokenHeader, originOf } = require('./service-client');

/**
 * Minimal reverse proxy from the gateway to an isolated agent service.
 *
 * The gateway is the only browser-facing origin; agent read endpoints
 * (/api/agent, /api/coder) are served by separate internal services. This
 * forwards the request (method, path + query, JSON body) and returns the
 * response. All agent endpoints speak JSON, so the body is reconstructed from
 * the already-parsed req.body.
 *
 * The S2S auth (Google OIDC ID token minting for Cloud Run IAM) lives in
 * service-client.js so the proxy and the gateway's own S2S calls share one
 * implementation.
 */

// opts.rewrite = { from, to } rewrites the leading path segment (e.g. mount at
// /api/org but the target service serves /api/v1). opts.forwardUserAuth carries
// the caller's bearer to a service that does its OWN token auth (the org service
// verifies the Firebase ID token): the user token rides in
// X-Forwarded-Authorization while Authorization holds the S2S OIDC token that
// Cloud Run's IAM check requires.
function createProxy(baseUrl, opts = {}) {
  const audience = originOf(baseUrl);

  return async function proxy(req, res) {
    let path = req.originalUrl;
    if (opts.rewrite) path = path.replace(opts.rewrite.from, opts.rewrite.to);
    const target = `${baseUrl}${path}`;
    const headers = {};
    const init = { method: req.method, headers };

    const hasBody = req.method !== 'GET' && req.method !== 'HEAD';
    if (hasBody && req.body && Object.keys(req.body).length > 0) {
      init.body = JSON.stringify(req.body);
      headers['content-type'] = 'application/json';
    }

    const incomingAuth = (req.get && req.get('authorization')) || '';
    if (opts.forwardUserAuth && incomingAuth) {
      headers['x-forwarded-authorization'] = incomingAuth;
    }

    if (CONFIG.MESSAGING_MODE === 'pubsub') {
      try {
        const authorization = await idTokenHeader(audience);
        if (authorization) headers.authorization = authorization;
      } catch (err) {
        log.error(`gateway proxy could not mint ID token for ${audience}: ${err && err.message ? err.message : err}`);
      }
    } else if (opts.forwardUserAuth && incomingAuth) {
      // Local/direct mode: no S2S OIDC — pass the user's bearer straight through.
      headers.authorization = incomingAuth;
    }

    // Trusted headers the gateway derives server-side (e.g. resolved org
    // membership). Safe: this service is reachable only via the gateway SA
    // (Cloud Run IAM) and `headers` starts from {} — client-supplied values are
    // never copied through, so these cannot be spoofed by the browser.
    if (typeof opts.injectHeaders === 'function') {
      const injected = opts.injectHeaders(req) || {};
      for (const [key, value] of Object.entries(injected)) {
        if (value !== undefined && value !== null && value !== '') headers[key.toLowerCase()] = String(value);
      }
    }

    try {
      const resp = await fetch(target, init);
      const text = await resp.text();
      res.status(resp.status);
      const contentType = resp.headers.get('content-type');
      if (contentType) res.set('content-type', contentType);
      res.send(text);
    } catch (err) {
      const message = err && err.message ? err.message : String(err);
      log.error(`gateway proxy ${req.method} ${target} failed: ${message}`);
      res.status(502).json({ error: `Agent service unavailable: ${message}` });
    }
  };
}

module.exports = { createProxy };

File: ai_fleet/services/gateway/proxy.py
"""Minimal reverse proxy from the gateway to an isolated agent service
(port of services/gateway/src/proxy.js).

The gateway is the only browser-facing origin; agent endpoints (/api/agent,
/api/coder) are served by separate service processes. This forwards the request
verbatim (method, path + query, JSON body) to the target service and streams the
response status/body back. A network failure surfaces as 502 with a clear
message instead of a hung request.
"""

from __future__ import annotations

import httpx
from starlette.requests import Request
from starlette.responses import Response

from ai_fleet import logger


def create_proxy(base_url: str):
    async def proxy(request: Request) -> Response:
        # request.url includes path + query; forward it verbatim onto the target base.
        target = f"{base_url}{request.url.path}"
        if request.url.query:
            target = f"{target}?{request.url.query}"
        headers = {}
        body = None
        if request.method not in ("GET", "HEAD"):
            raw = await request.body()
            if raw:
                body = raw
                headers["content-type"] = "application/json"
        try:
            async with httpx.AsyncClient(timeout=None) as client:
                resp = await client.request(request.method, target, headers=headers, content=body)
            content_type = resp.headers.get("content-type")
            return Response(
                content=resp.content,
                status_code=resp.status_code,
                media_type=content_type,
            )
        except httpx.HTTPError as err:
            message = str(err) or repr(err)
            logger.error(f"gateway proxy {request.method} {target} failed: {message}")
            return Response(
                content=f'{{"error": "Agent service unavailable: {message}"}}',
                status_code=502,
                media_type="application/json",
            )

    return proxy
